utils/users/src/users.ts

```diff
@@ -0,0 +1,209 @@
+      .on("error", (error) => reject(error))
+  })
+}
+
+/**
+ * Get an access token from keycloak for the SIMS Service account.
+ *
+ * @return {*}
+ */
+async function getKeycloakToken() {
+  const response = await axios.post(
+    KEYCLOAK_API_TOKEN_URL,
+    {
+      client_id: SIMS_KEYCLOAK_API_CLIENT_ID,
+      client_secret: SIMS_KEYCLOAK_API_CLIENT_PASSWORD,
+      grant_type: "client_credentials",
+    },
+    { headers: { "Content-Type": "application/x-www-form-urlencoded" } }
+  )
+
+  return response.data.access_token
+}
+
+/**
+ * Get keycloak data by email
+ *
+ * @param {IUser} user
+ * @return {*}
+ */
+async function getKeycloakIDIR_UUID_ByEmail(user: IUser) {
+  const response = await axios.get(
+    `${KEYCLOAK_API_USER_URL}/${environment}/idir/users?email=${user.EMAIL_ADDRESS}`,
+    { headers: { Authorization: `Bearer ${token}` } }
+  )
+  if (response.data.data.length > 0) {
+    const keycloakUserData = response.data.data[0]
+    user.keycloakData = keycloakUserData
+
+    return user
+  }
+}
+
+/**
+ * Get keycloak data by name
+ *
+ * @param {IUser} user
+ * @return {*}
+ */
+async function getKeycloakIDIR_UUID_ByName(user: IUser) {
+  const response = await axios.get(
+    `${KEYCLOAK_API_USER_URL}/${environment}/idir/users?firstName=${user.FIRST_NAME}&lastName=${user.LAST_NAME}`,
+    { headers: { Authorization: `Bearer ${token}` } }
+  )
+  if (response.data.data.length > 0) {
+    const keycloakUserData = response.data.data[0]
+    user.keycloakData = keycloakUserData
+
+    return user
+  }
+}
+
+/**
+ * Get keycloak data by email or name
+ *
+ * @param {IUser} user
+ * @return {*}
+ */
+async function getKeycloakData(user: IUser) {
+  let keycloakUser: IUser | undefined
+
+  if (user.EMAIL_ADDRESS) {
+    try {
+      keycloakUser = await getKeycloakIDIR_UUID_ByEmail(user)
+    } catch (e: any) {
+      if (e.response.data.err === "token expired") {
+        console.log("token expired")
+        token = await getKeycloakToken()
+        keycloakUser = await getKeycloakIDIR_UUID_ByEmail(user)
+      }
+    }
+    if (keycloakUser) {
+      return keycloakUser
+    }
+  }
+
+  if (user.FIRST_NAME && user.LAST_NAME) {
+    try {
+      keycloakUser = await getKeycloakIDIR_UUID_ByName(user)
+    } catch (e: any) {
+      if (e.response.data.err === "token expired") {
+        console.log("token expired")
+        token = await getKeycloakToken()
+        keycloakUser = await getKeycloakIDIR_UUID_ByName(user)
+      }
+    }
+    if (keycloakUser) {
+      return keycloakUser
+    }
+  }
+}
+
+/**
+ * Get all users from keycloak
+ *
+ * @export
+ * @param {IUser[]} users
+ * @return {*}
+ */
+export async function getAllUsers(users: IUser[]) {
+  const idirUsers: IUser[] = []
+
+  for (const user of users) {
+    const keycloakUser = await getKeycloakData(user)
+    if (keycloakUser) {
+      console.log("keycloakUser", keycloakUser)
+      idirUsers.push(keycloakUser)
+      fs.appendFileSync(
```

Comment:
Could we not just push all the keycloak users into an array and write the entire array to file all at once?

Reply:
I do this at line 203 for the file `keycloakUsers.json` I initially did it in steps as to help debug breaking points.